server/src/main/java/org/elasticsearch/cluster/coordination/PreVoteCollector.java

```diff
@@ -0,0 +1,192 @@
+                        handlePreVoteResponse(response, n);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (exp.getRootCause() instanceof CoordinationStateRejectedException) {
+                            logger.debug("{} failed: {}", this, exp.getRootCause().getMessage());
+                        } else {
+                            logger.debug(new ParameterizedMessage("{} failed", this), exp);
+                        }
+                    }
+
+                    @Override
+                    public String executor() {
+                        return Names.GENERIC;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "TransportResponseHandler{" + PreVoteCollector.this + ", node=" + n + '}';
+                    }
+                }));
+        }
+
+        private void handlePreVoteResponse(final PreVoteResponse response, final DiscoveryNode sender) {
+            if (isClosed.get()) {
+                logger.debug("{} is closed, ignoring {} from {}", this, response, sender);
+                return;
+            }
+
+            // TODO the response carries the sender's current term. If an election starts then it should be in a higher term.
+
+            if (response.getLastAcceptedTerm() > clusterState.term()
+                || (response.getLastAcceptedTerm() == clusterState.term()
+                && response.getLastAcceptedVersion() > clusterState.version())) {
+                logger.debug("{} ignoring {} from {} as it is fresher", this, response, sender);
+                return;
+            }
+
+            preVotesReceived.add(sender);
+            final VoteCollection voteCollection = new VoteCollection();
+            preVotesReceived.forEach(voteCollection::addVote);
+
+            if (isElectionQuorum(voteCollection, clusterState) == false) {
+                logger.debug("{} added {} from {}, no quorum yet", this, response, sender);
+                return;
+            }
+
+            if (electionStarted.compareAndSet(false, true) == false) {
+                logger.debug("{} added {} from {} but election has already started", this, response, sender);
+                return;
+            }
+
+            logger.debug("{} added {} from {}, starting election", this, response, sender);
+            startElection.run();
+        }
+
+        @Override
+        public String toString() {
+            return "PreVotingRound{" +
+                "preVotesReceived=" + preVotesReceived +
+                ", electionStarted=" + electionStarted +
+                ", preVoteRequest=" + preVoteRequest +
+                ", isClosed=" + isClosed +
+                '}';
+        }
+
+        @Override
+        public void close() {
+            final boolean isClosedChanged = isClosed.compareAndSet(false, true);
```

Comment:
maybe name this `isNotAlreadyClosed`

Reply:
Done in 4f61eff3467d29231b7cf4fb53d7a23c585ae9da